src/modules-load/modules-load.c

```diff
@@ -163,34 +179,42 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 log_warning_errno(r, "Failed to parse kernel command line, ignoring: %m");
 
-        r = module_setup_context(&ctx);
-        if (r < 0)
-                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+        _cleanup_close_pair_ int seq[2] = EBADF_PAIR;
+        const int max_tasks = MAX(MIN(sysconf(_SC_NPROCESSORS_ONLN) / 2, 10), 2);
+        _cleanup_free_ pthread_t *threads = malloc(sizeof(pthread_t) * max_tasks);
+        if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, seq) == -1)
+                log_error_errno(errno, "socketpair");
 
-        r = 0;
+        // Create worker threads
+        for (int i = 0; i < max_tasks; ++i)
+                if (pthread_create(&threads[i], NULL, worker_thread, &seq[1]) != 0)
+                        log_error_errno(errno, "pthread_create");
 
+        r = 0;
         if (argc > optind) {
                 for (int i = optind; i < argc; i++)
-                        RET_GATHER(r, apply_file(ctx, argv[i], false));
-
+                        RET_GATHER(r, apply_file(seq, argv[i], false));
         } else {
                 _cleanup_strv_free_ char **files = NULL;
-
-                STRV_FOREACH(i, arg_proc_cmdline_modules) {
-                        k = module_load_and_warn(ctx, *i, true);
-                        if (k == -ENOENT)
-                                continue;
-                        RET_GATHER(r, k);
-                }
+                STRV_FOREACH(i, arg_proc_cmdline_modules)
+                        if (send(seq[0], *i, strlen(*i), 0))
```

Comment:
I think the call to send() should have a function of its own, that logs about send errors (but generally continues)

Reply:
(i mean, send() will fail if you pass an overly long kmod name, because that will hit the datagram size limit. That this fails is expected and ok, and such long kmod names don't make any sense, hence we should warn, but continue)